src/MongoDB.Driver.Core/Core/NativeLibraryLoader/PinnedBuffer.cs

```diff
@@ -15,24 +15,40 @@
 
 using System;
 using System.Runtime.InteropServices;
+using MongoDB.Driver.Core.Misc;
 
 namespace MongoDB.Driver.Core.NativeLibraryLoader
 {
     internal class PinnedBuffer : IDisposable
     {
+        private readonly byte[] _bytes;
         private GCHandle _handle;
-        private readonly IntPtr _intPtr;
+        private IntPtr _intPtr;
+        private int _offset;
 
         public PinnedBuffer(byte[] bytes, int offset)
         {
+            _bytes = Ensure.IsNotNull(bytes, nameof(bytes));
             // The array must be pinned by using a GCHandle before it is passed to UnsafeAddrOfPinnedArrayElement.
             // For maximum performance, this method does not validate the array passed to it; this can result in unexpected behavior.
-            _handle = GCHandle.Alloc(bytes, GCHandleType.Pinned);
-            _intPtr = Marshal.UnsafeAddrOfPinnedArrayElement(bytes, offset);
+            _handle = GCHandle.Alloc(_bytes, GCHandleType.Pinned);
+            _offset = offset;
+
+            RefreshIntPtr();
         }
 
         public IntPtr IntPtr => _intPtr;
 
+        public int Offset
+        {
+            get => _offset;
+            set
+            {
+                _offset = value;
+                RefreshIntPtr();
```

Comment:
Ah, we need to be able to mutate the offset now :\ Alas for immutability!

Reply:
That being said, your solution to this need is neat :)

Reply:
I think this is a bad change.

PinnedBuffer was immutable and now it isn't.

Reply:
Instead of **assigning** to `Offset` and changing it, we could have a method instead:

    public IntPtr GetIntPtrToOffset(int offset);

Reply:
in this case, we will need to remove offset property because it looks strange that we have immutable offset in the class itself, but if we want to get memory pointer for a specific(not initial) offset(which means exactly the same business idea as the initial(immutable) one), we will need to call a method. I will think about it

Reply:
If this change was local to ZStandard I wouldn't mind so much.

But this class is a general purpose class and it would be desirable that we not make it mutable.